### DISCUSSION TITLE
Next js state delay issue

### DISCUSSION THREAD

**Discussion by NikCrmr** [neuefische/questions#831] [Web] [answered]
We have an web-app called mood-tracker in the dolphin team.
There is one bug reoccuring, maybe it's cause it's getting warmer and bug's get more active anyway or it's an issue with the code.

We have mood-entries and they are visualized in the BarChart component. the parent component holding all the logic is the Dashboard component. In the bard chart you have three filter options to visualize the data. Visualized are 5 emotions-users-entries either the 7 days of it or the whole users database. Also there is the option to just click on one emotion, for example the yellow bar for enjoyment and filter only the list and entries for just one emotion.

The Bug:

The bug occurs when...
-  you click on one single emotion >> then you get the list of just one emotion
- then you deselect it by clicking again >> you get all entries /days or all (what was selected)
Also the clicked emotion is white highlighted with a border. 
- Then you click another emotion for example sadness and now you should see the list of sadness entries, but there often nothing happens and there you have to click again on the emotion. you have to double click or click around til the emotions then are displayed.

That's the bug. 

here you can check out the bug. You need to go to the entries section. 
https://mood-diary-git-v2-barchart-bugfix-mathis-santos-projects.vercel.app/inddd
here is the feature branch for the bugfix
https://github.com/MathiSans/capstone_mood-tracker/tree/v2/barchart-bugfix

This is the **Barchart.js** the Child

```
`      <Styled.BarChartContainer>
        {visualizedData &&
          visualizedData.experiences.map(
            ({ index, count, color, experience }) => (
              <Styled.SingleBar
                key={index}
                color={color}
                experience={experience}
                barHeight={Math.floor((count / totalCount) * 100)}
                onClick={(event) => {
                  handleExperienceClick(event, experience);
                }}
                isClicked={
                  singleExperienceList && clickedExperience === experience
                }
              />
            )
          )}
      </Styled.BarChartContainer>`
```

This is the parent component the **Dashboard.js**

```
const handleExperienceClick = (event, experience) => {
    const targetValues = Object.values(event.target);
    const experienceBarChart = targetValues[1].experience;
    setClickedExperience(experienceBarChart);
    setTargetExperience(experienceBarChart);
    setSingleExperienceList(!singleExperienceList);

    console.log(targetExperience, experienceBarChart, experience);

    if (experience === targetExperience) {
      setSingleExperienceList(!singleExperienceList);
      // setClickedExperience(null);

      console.log("singleExperienceList", singleExperienceList);
    }
    if (experience !== targetExperience) {
      console.log("experience is not experience");
    }
  };
```

This is as link to the Barchart.js

https://github.com/MathiSans/capstone_mood-tracker/blob/v2/barchart-bugfix/components/Dashboard/Tiles/BarChartTile/BarChartTile.js

this to the Dashboard.js

https://github.com/MathiSans/capstone_mood-tracker/blob/v2/barchart-bugfix/components/Dashboard/Dashboard.js

**This is the how the components is aligned with it's parent components:**

<img width="521" alt="Bildschirmfoto 2024-05-03 um 10 00 07" src="https://github.com/neuefische/questions/assets/148204632/2c3146ae-9a3e-4806-ab86-8a30f8ffbda1">

**Comment by onemanwenttomow** [comment, marked as answer]
Ok, so i had a little look at your code. 

First of all i would like you to work on getting rid of all of the warnings and errors in the console. Most of them involve not using $ with props to styled components. There were also issues with a `key` being undefined. 

In terms of fixing the "double click" issue, i think it makes sense to do the following. 

In BarChartTitle i don't think there is any need to pass the `event` object when calling `handleExperienceClick`. The information we are trying to find on the event, is already available to us in `experience`.

``` js
<Styled.BarChartContainer>
  {visualizedData &&
    visualizedData.experiences.map(({ count, color, experience }) => (
      <Styled.SingleBar
        key={experience}
        $color={color}
        $experience={experience}
        $barHeight={Math.floor((count / totalCount) * 100)}
        onClick={() => handleExperienceClick(experience)}
        $isClicked={
          singleExperienceList && clickedExperience === experience
        }
      />
    ))}
</Styled.BarChartContainer>
```

We can then simplify `handleExperienceClick` to the following.

``` js
const handleExperienceClick = (experience) => {
  setClickedExperience(experience);
  setTargetExperience(experience);
  setSingleExperienceList(
    experience !== clickedExperience || !singleExperienceList
  );
};

```

**Reply by NikCrmr** [reply]
Thank you Peter!
